### PR TITLE
fix high cpu load

### PR DESCRIPTION
**Pull request by qingtao**
I used this packege when writing an mqtt program and ran into problems with high cpu load. I forked this repository and increased wait times on retries based on similar processing in go-cloud, which can significantly reduce cpu load, will you incorporate this change if needed?

### CODE REVIEW

File: mqtt_test.go [hunk 1]
```diff
@@ -38,11 +40,15 @@ type harness struct {
 	Publisher
 }
 
-func newHarness(ctx context.Context, t *testing.T) (drivertest.Harness, error) {
+type _logger interface {
```
Comment (filinvadim) [inline mqtt_test.go:43]: Sorry, please remove underscore :)

File: pubsub.go [hunk 2]
```diff
@@ -172,38 +175,41 @@ func (t *topic) SendBatch(ctx context.Context, msgs []*driver.Message) error {
 	if t == nil || t.conn == nil {
 		return errConnRequired
 	}
+	t.mu.Lock()
```
Comment (filinvadim) [inline pubsub.go:178]: Sorry, i don't get it. You are synchronized this func completely. There is no point in goroutines now

File: pubsub.go [hunk 3]
```diff
@@ -281,33 +289,62 @@ func (s *subscription) ReceiveBatch(ctx context.Context, maxMessages int) (dms [
...
+	s.mu.Unlock()
+	if retry {
+		if s.opts.WaitTime <= 0 {
+			return nil, nil
+		}
+		timer := time.NewTimer(s.opts.WaitTime)
+		// fmt.Println("timer")
+		select {
+		case <-timer.C:
+		case <-ctx.Done():
```
Comment (filinvadim) [inline pubsub.go:304]: https://dave.cheney.net/2017/08/20/context-isnt-for-cancellation

File: pubsub.go [hunk 4]
```diff
@@ -319,9 +356,16 @@ func (s *subscription) SendAcks(ctx context.Context, ids []driver.AckID) error {
 		return nil
 	}
 
+	s.mu.Lock()
```
Comment (filinvadim) [inline pubsub.go:359]: Same here. Func now is synced

**Comment by qingtao** [comment]
> Thank you @qingtao! Excellent work! I've left just couple comments: of course synchronizing everything will reduce cpu load.

You're probably right, but I don't agree with you about context not being used for cancellation. go-cloud receives messages by creating new goroutine, and if there is no message coming, without a wait time and a mutex, it will generate a lot of goroutine, which is the cause of excessive cpu load when idle. Thank you for pointing out the problem of this PR, and look forward to your better solution.

**Comment by filinvadim** [comment]
Thanks! Very convincing! Could you remove unused and excessive go routines that were written by me? Except this - LGTM.

**Comment by qingtao** [comment]
[google/go-cloud#2103](https://github.com/google/go-cloud/issues/2103)

What I'm saying should relate to this, and refer to how it was fixed in rabbitpubsub.

**Pull request merged**